src/esp/gfx/replay/Player.cpp

```diff
@@ -223,27 +221,78 @@ void Player::applyKeyframe(const Keyframe& keyframe) {
     const auto& instanceKey = pair.first;
     CORRADE_INTERNAL_ASSERT(createdInstances_.count(instanceKey) == 0);
     createdInstances_[instanceKey] = node;
+    creationInfos_[instanceKey] = adjustedCreation;
   }
 
-  if (!frameCleared) {
+  // TODO: Move the code below into renderer-specific implementations.
+  bool isClassicReplayRenderer =
+      dynamic_cast<AbstractSceneGraphPlayerImplementation*>(
+          implementation_.get()) != nullptr;
+  if (isClassicReplayRenderer) {
     for (const auto& deletionInstanceKey : keyframe.deletions) {
       const auto& it = createdInstances_.find(deletionInstanceKey);
       if (it == createdInstances_.end()) {
-        // missing instance for this key, probably due to a failed instance
-        // creation
+        // Missing instance for this key due to a failed instance creation
         continue;
       }
 
       implementation_->deleteAssetInstance(it->second);
       createdInstances_.erase(deletionInstanceKey);
     }
+    // The batch renderer can only clear the scene entirely; it cannot delete
+    // individual objects. To process deletions, all instances are deleted,
+    // remaining instances are re-created and latest transform updates are
+    // re-applied.
+  } else if (keyframe.deletions.size() > 0) {
+    // Cache latest transforms
+    std::unordered_map<RenderAssetInstanceKey, Mn::Matrix4> latestTransforms{};
```

Comment:
Let's get out of the habit of per-frame dynamic allocations. Make this a class member and then clear and re-use it here.